File: leetsync/2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level/minimum-number-of-operations-to-sort-a-binary-tree-by-level.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type queue struct {
	node []*TreeNode
}

func NewQueue() *queue {
	return &queue{}
}

func (q *queue) Size() int {
	return len(q.node)
}

func (q *queue) Pop() *TreeNode {
	v := q.node[0]
	q.node = q.node[1:]
	return v
}

func (q *queue) Push(v *TreeNode) {
	q.node = append(q.node, v)
}

func minimumOperations(root *TreeNode) int {
	totalSwap := 0
	levelValues := make([]int, 0, 8)
	q := NewQueue()
	q.Push(root)
	for q.Size() > 0 {
		levelValues = levelValues[:0]
		for qSize := q.Size(); qSize > 0; qSize-- {
			node := q.Pop()
			levelValues = append(levelValues, node.Val)
			if node.Left != nil {
				q.Push(node.Left)
			}
			if node.Right != nil {
				q.Push(node.Right)
			}
		}
		totalSwap += getMinSwap(levelValues)
	}
	return totalSwap
}

func getMinSwap(nums []int) int {
	swaps := 0
	sortedNums := make([]int, len(nums))
	copy(sortedNums, nums)
	sort.Ints(sortedNums)
	pos := make(map[int]int, len(nums))
	for i, v := range nums {
		pos[v] = i
	}
	for i, v := range nums {
		if nums[i] != sortedNums[i] {
			swaps++
			needPos := pos[sortedNums[i]]
			pos[v] = needPos
			nums[i], nums[needPos] = nums[needPos], nums[i]
		}
	}

	return swaps
}

File: leetsync/2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level/minimum-number-of-operations-to-sort-a-binary-tree-by-level.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int minimumOperations(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int result = 0;
        while (!q.isEmpty()) {
            int n = q.size();
            List<Integer> nums = new ArrayList<>();
            while (n-- > 0) {
                TreeNode node = q.poll();
                nums.add(node.val);
                if (node.left != null) {
                    q.add(node.left);
                }
                if (node.right != null) {
                    q.add(node.right);
                }
            }
            result += countMinSwapToSort(nums);
        }
        return result;
    }

    public int countMinSwapToSort(List<Integer> nums) {
        Map<Integer, Integer> map = new HashMap();
        int swaps = 0;
        List<Integer> sortNums = new ArrayList<>(nums);
        Collections.sort(sortNums);
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            map.put(nums.get(i), i);
        }
        for (int i = 0; i < n; i++) {
            if (sortNums.get(i).equals(nums.get(i))) {
                continue;
            }
            int currentIndex = map.get(sortNums.get(i));
            map.put(nums.get(i), currentIndex);
            map.put(nums.get(currentIndex), i);
            Collections.swap(nums, currentIndex, i);
            swaps++;
        }
        return swaps;
    }
}